### ISSUE TITLE
[BUG] - Conda-store fails to build valid environment yaml

### ISSUE THREAD

**Issue by kcpevey** [conda-incubator/conda-store#712]
### Describe the bug

I am trying to build an environment that has a git repo in the `pip` section of the environment.yml (see below). 

When I try to build this environment on a Nebari deployment `v2023.11.1`, the build fails. If I copy/paste the same environment spec in a Nebari deployment from `v2023.7.2.dev62+g9179295`, the build will pass. It will also build on my local machine. 

What is also interesting is that if you swap that pip package for a different one, it will fail with a new error. I believe what I'm seeing are indications that conda-store isn't parsing the yaml spec properly (though I could be wrong). 

### Expected behavior

I expect to be able to build an environment with a git repo in the pip section. 

### How to Reproduce the problem?

```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - param @ git+https://github.com/holoviz/param.git@main
  - ipykernel

```

### Output

```
starting build of conda environment 2023-12-18 22:27:52.221978 UTC
Traceback (most recent call last):
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/toml/file.py", line 32, in read
    return super(TOMLFile, self).read()
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/toml_file.py", line 20, in read
    return loads(f.read())
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/api.py", line 34, in loads
    return parse(string)
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/api.py", line 51, in parse
    return Parser(string).parse()
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 160, in parse
    key, value = self._parse_table()
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 1176, in _parse_table
    key_next, table_next = self._parse_table(name, table)
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 1152, in _parse_table
    item = self._parse_item()
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 306, in _parse_item
    return self._parse_key_value(True)
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 397, in _parse_key_value
    val = self._parse_value()
  File "/opt/conda/envs/conda-store-server/lib/python3.10/site-packages/conda_lock/_vendor/poetry/core/_vendor/tomlkit/parser.py", line 669, in _parse_value
    raise self.parse_error(UnexpectedCharError, c)
tomlkit.exceptions.UnexpectedCharError: Unexpected character: '.' at line 182 col 18
```

### Versions and dependencies used.

I'm not sure I have the powers to view the conda-store version that is actually deployed.

**Comment by ericdatakelly** [comment]
Adding `@tag` worked for me in a private gitlab repo.

```yaml
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - <package name> @
        git+https://<username>/:<PAT>@gitlab.....git@v0.3.0
  - ipykernel
```

But using `@main` failed.

**Comment by nkaretnikov** [comment]
Status update: I've followed up with Kim offline with some ideas to try.

**Comment by kcpevey** [comment]
Some additional findings:

I have a private package `adversarial-robustness-toolbox-jatic` which has an `install_requires` dependency on another private tool `jatic-toolbox`. If I add them both with PATs to my yaml file, I can solve them on my machine (maybe because of my ssh keys???). But on conda-store I get these results:

**art-jatic only**
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - adversarial-robustness-toolbox-jatic @
        git+https://{username}:{PAT}@gitlab...git@0.1.0
  - ipykernel
```

Fails with 
```
starting build of conda environment 2023-12-20 14:15:41.561800 UTC
Cloning into '/tmp/pypoetry-git-jatic-toolbox4sc58huu'...
fatal: could not read Username for 'https://gitlab.jatic.net': No such device or address
```

**combined yaml** 
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - jatic-toolbox @
        git+https://{username}:{PAT}@gitlab...git@v0.2.0
      - adversarial-robustness-toolbox-jatic @
        git+https://{username}:{PAT}@gitlab...git@0.1.0
  - ipykernel
 ```
Fails with the same error as above

**jatic-toolbox only**
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - jatic-toolbox @
        git+https://{username}:{PAT}@gitlab...git@v0.2.0
  - ipykernel
```
Solves on conda-store properly.

**Comment by kcpevey** [comment]
Some additional findings:

I have a private package `adversarial-robustness-toolbox-jatic` which has an `install_requires` dependency on another private tool `jatic-toolbox`. If I add them both with PATs to my yaml file, I can solve them on my machine (maybe because of my ssh keys???). But on conda-store I get these results:

**art-jatic only**
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - adversarial-robustness-toolbox-jatic @
        git+https://{username}:{PAT}@gitlab...git@0.1.0
  - ipykernel
```

Fails with 
```
starting build of conda environment 2023-12-20 14:15:41.561800 UTC
Cloning into '/tmp/pypoetry-git-jatic-toolbox4sc58huu'...
fatal: could not read Username for 'https://gitlab...net': No such device or address
```

**combined yaml** 
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - jatic-toolbox @
        git+https://{username}:{PAT}@gitlab...git@v0.2.0
      - adversarial-robustness-toolbox-jatic @
        git+https://{username}:{PAT}@gitlab...git@0.1.0
  - ipykernel
 ```
Fails with the same error as above

**jatic-toolbox only**
```
channels:
  - conda-forge
dependencies:
  - python = 3.10
  - pip
  - pip:
      - jatic-toolbox @
        git+https://{username}:{PAT}@gitlab...git@v0.2.0
  - ipykernel
```
Solves on conda-store properly.

**Comment by dharhas** [comment]
This issue conflates 2 issues. One is that param isn't installing even though it is a public repo and the second is an issue related to private repos and passing credentials through. Can we split them please.

**Comment by kcpevey** [comment]
Apologies, I got a little confused and conflated some things. 

If you add a private package to your environment spec, it must be done in the proper format ([documentation for that is being tracked here](https://github.com/conda-incubator/conda-store/issues/497)). 

Where I had overlapping problems is when I was trying to add a private package which had private subdependencies. Specifically, the private package that I added directly to the yaml had a direct dependency on a private git repo specified as ssh syntax. We cannot currently handle this in conda-store. Even adding this private subdependency into your yaml spec with a token will not avoid a failure. 

I will close this issue as high level private package syntax is being added to the docs in [another issue](https://github.com/conda-incubator/conda-store/issues/497) and conda-store management of secrets is a major refactor that we are not targeting at this time.
